system/jlib/jmetrics.cpp

```diff
@@ -50,28 +60,39 @@ void MetricsManager::init(IPropertyTree *pMetricsTree)
 }
 
 
-void MetricsManager::addMetric(const std::shared_ptr<IMetric> &pMetric)
+bool MetricsManager::addMetric(const std::shared_ptr<IMetric> &pMetric)
 {
+    bool rc = false;
     std::unique_lock<std::mutex> lock(metricVectorMutex);
-    auto it = metrics.find(pMetric->queryName());
+    std::string name = pMetric->queryName();
+    auto metaData = pMetric->queryMetaData();
+    for (auto &metaDataIt: metaData)
+    {
+        name.append(".").append(metaDataIt.value);
+    }
+
+    auto it = metrics.find(name);
     if (it == metrics.end())
     {
-        metrics.insert({pMetric->queryName(), pMetric});
+        metrics.insert({name, pMetric});
+        rc = true;
     }
     else
     {
-        //If there is a match only report an error if the metric has not been destroyed in the meantime
+        rc = false;
+        // If there is a match only report an error if the metric has not been destroyed in the meantime
         auto match = it->second.lock();
         if (match)
         {
 #ifdef _DEBUG
-            throw MakeStringException(MSGAUD_operator, "addMetric - Attempted to add duplicate named metric with name '%s'", pMetric->queryName().c_str());
+            throw MakeStringException(MSGAUD_operator, "addMetric - Attempted to add duplicate named metric with name '%s'", name.c_str());
 #else
-            OERRLOG("addMetric - Adding a duplicate named metric '%s', old metric replaced", pMetric->queryName().c_str());
+            OERRLOG("addMetric - Adding a duplicate named metric '%s', old metric replaced", name.c_str());
 #endif
         }
         it->second = pMetric;
```

Comment:
If it does return a value, then it should be clearly defined.  In this case it should be the same as whether an exception would not be thrown.  That means it should also set it to true if match is non null (otherwise your regression tests would fail in release) - and should be extended to cover this case.

Reply:
Added a comment that the return value indicates whether the name was unique when the metric was added.